source/beamline/CLS/CLSBeamline.cpp

```diff
@@ -1,12 +1,227 @@
+
+	beamlineStatus_ = 0;
+	beamlineShutters_ = 0;
+	beamlineValves_ = 0;
+
+	beamOnAction_ = 0;
+	beamOffAction_ = 0;
+
+	requiredControls_ = new AMControlSet(this);
+	requiredDetector_ = new AMDetectorSet(this);
+
+	connect(CLSStorageRing::sr1(), SIGNAL(beamAvaliability(bool)), this, SLOT(updateBeamStatus()));
+	connect(requiredControls_, SIGNAL(connected(bool)), this, SLOT(onBeamlineComponentConnected()));
+	connect(requiredDetector_, SIGNAL(connected(bool)), this, SLOT(onBeamlineComponentConnected()));
+}
+
+CLSBeamline::~CLSBeamline() { }
+
+/// ==================== public slots =======================
+
+bool CLSBeamline::isConnected() const
+{
+	return	(requiredControls_->isEmpty() || requiredControls_->isConnected())
+		 && (requiredDetector_->isEmpty() || requiredDetector_->isConnnected());
+}
+
+CLSBeamlineStatus* CLSBeamline::beamlineStatus() const
+{
+	return beamlineStatus_;
+}
+
+/// ==================== public slots =======================
+void CLSBeamline::onTurningBeamOnRequested(){
+	if(beamOnAction_)
+		return;
 
+	beamOnAction_ = createBeamOnActions();
+	if (beamOnAction_) {
+		connect(beamOnAction_, SIGNAL(succeeded()), this, SLOT(onBeamOnActionFinished()));
+		connect(beamOnAction_, SIGNAL(failed()), this, SLOT(onBeamOnActionFinished()));
+
+		beamOnAction_->start();
+	}
+}
+
+void CLSBeamline::onTurningBeamOffRequested(){
+	if(beamOffAction_)
+		return;
+
+	beamOffAction_ = createBeamOffActions();
+	if (beamOffAction_) {
+		connect(beamOffAction_, SIGNAL(succeeded()), this, SLOT(onBeamOffActionFinished()));
+		connect(beamOffAction_, SIGNAL(failed()), this, SLOT(onBeamOffActionFinished()));
+		beamOffAction_->start();
+	}
+}
+
+
+/// ==================== protected slots =======================
+
+void CLSBeamline::onBeamlineComponentConnected()
+{
+	bool currentConnectedStatus = isConnected();
+	if (wasConnected_ != currentConnectedStatus) {
+		wasConnected_ = currentConnectedStatus;
+		emit connected(wasConnected_);
+	}
 }
 
-CLSBeamline::~CLSBeamline()
+void CLSBeamline::updateBeamStatus()
 {
+	bool beamOn = CLSStorageRing::sr1()->beamAvailable();
+	if (beamlineStatus_) {
+		beamOn &= beamlineStatus_->isOn();
+	}
+
+	if (wasBeamOn_ != beamOn) {
+		wasBeamOn_ = beamOn;
+		emit beamAvaliabilityChanged(beamOn);
+	}
+}
+
+void CLSBeamline::onBeamOnActionFinished(){
+	disconnect( beamOnAction_, 0, this, 0 );
+
+	beamOnAction_->deleteLater();
+	beamOnAction_ = 0; //NULL
+}
+
+void CLSBeamline::onBeamOnActionFailed(){
+	AMErrorMon::error(this, CLSBEAMLINE_ERR_BEAM_ON_FAILED, QString("Failed to execute the beam on actions with message: %1.").arg(beamOnAction_->failureMessage()), true);
+	onBeamOnActionFinished();
+}
+
+void CLSBeamline::onBeamOffActionFinished(){
+	disconnect( beamOffAction_, 0, this, 0 );
+
+	beamOffAction_->deleteLater();
+	beamOffAction_ = 0; //NULL
+}
+
+/// ==================== protected methods =======================
+
+void CLSBeamline::createBeamlineStatus(CLSShutters *shutters, CLSValves *valves)
```

Comment:
It looks like the ring current could also be incorporated eventually, based on the update beam status method above.